File: src/msg_simix/host.c
#include "private.h"
#include "xbt/sysdep.h"
#include "xbt/log.h"

/** \defgroup m_host_management Management functions of Hosts
 *  \brief This section describes the host structure of MSG
 * 
 *     \htmlonly <!-- DOXYGEN_NAVBAR_LABEL="Hosts" --> \endhtmlonly
 * (#m_host_t) and the functions for managing it.
 *  
 *  A <em>location</em> (or <em>host</em>) is any possible place where
 *  a process may run. Thus it may be represented as a
 *  <em>physical resource with computing capabilities</em>, some
 *  <em>mailboxes</em> to enable running process to communicate with
 *  remote ones, and some <em>private data</em> that can be only
 *  accessed by local process.
 *  \see m_host_t
 */

/********************************* Host **************************************/
m_host_t __MSG_host_create(const char *name,
			 void *workstation,
			 void *data)
{
  m_host_t host = xbt_new0(s_m_host_t,1);

  return host;
}

/** \ingroup m_host_management
 *
 * \brief Set the user data of a #m_host_t.
 *
 * This functions checks whether some data has already been associated to \a host 
   or not and attach \a data to \a host if it is possible.
 */
MSG_error_t MSG_host_set_data(m_host_t host, void *data)
{
  xbt_assert0((host!=NULL), "Invalid parameters");
  xbt_assert0((host->data == NULL), "Data already set");

  /* Assign data */
  host->data = data;

  return MSG_OK;
}

/** \ingroup m_host_management
 *
 * \brief Return the user data of a #m_host_t.
 *
 * This functions checks whether \a host is a valid pointer or not and return
   the user data associated to \a host if it is possible.
 */
void *MSG_host_get_data(m_host_t host)
{

  xbt_assert0((host != NULL), "Invalid parameters");

  /* Return data */
  return (host->data);
}

/** \ingroup m_host_management
 *
 * \brief Return the name of the #m_host_t.
 *
 * This functions checks whether \a host is a valid pointer or not and return
   its name.
 */
const char *MSG_host_get_name(m_host_t host)
{

  xbt_assert0((host != NULL) && (host->simdata != NULL), "Invalid parameters");

  /* Return data */
  return (host->name);
}

/** \ingroup m_host_management
 *
 * \brief Return the location on which the current process is executed.
 */
m_host_t MSG_host_self(void)
{
  return MSG_process_get_host(MSG_process_self());
}

/*
 * Real function for destroy a host.
 * MSG_host_destroy is just  a front_end that also removes it from 
 * msg_global->host
 */
void __MSG_host_destroy(m_host_t host)
{
  return;
}

/** \ingroup m_host_management
 * \brief Return the current number of #m_host_t.
 */
int MSG_get_host_number(void)
{
  return (xbt_fifo_size(msg_global->host));
}

/** \ingroup m_host_management
 * \brief Return a array of all the #m_host_t.
 */
m_host_t *MSG_get_host_table(void)
{
  return ((m_host_t *)xbt_fifo_to_array(msg_global->host));
}

/** \ingroup m_host_management
 * \brief Return the number of MSG tasks currently running on a
 * #m_host_t. The external load is not taken in account.
 */
int MSG_get_host_msgload(m_host_t h)
{
  xbt_assert0((h!= NULL), "Invalid parameters");
  xbt_assert0(0, "Not implemented yet");

  return(0);
/*   return(surf_workstation_resource->extension_public->get_load(h->simdata->host)); */
}

/** \ingroup m_host_management
 * \brief Return the speed of the processor (in flop/s), regardless of 
    the current load on the machine.
 */
double MSG_get_host_speed(m_host_t h)
{
	return 0.0;
}

/** \ingroup msg_gos_functions
 * \brief Determine if a host is available.
 *
 * \param h host to test
 */
int MSG_host_is_avail (m_host_t h)
{
	return 0;
}

File: src/java/simgrid/msg/Host.java
/*
 * Bindings to the MSG hosts
 *
 * Copyright 2006,2007,2010 The SimGrid Team           
 * All right reserved. 
 *
 * This program is free software; you can redistribute 
 * it and/or modify it under the terms of the license 
 *(GNU LGPL) which comes with this package. 
 *
 */  
package simgrid.msg;

/**
 * A host object represents a location (any possible place) where a process may run. 
 * Thus it is represented as a physical resource with computing capabilities, some 
 * mailboxes to enable running process to communicate with remote ones, and some private 
 * data that can be only accessed by local process. An instance of this class is always 
 * binded with the corresponding native host. All the native hosts are automatically created 
 * during the call of the method Msg.createEnvironment(). This method take as parameter a
 * platform file which describes all elements of the platform (host, link, root..).
 * You cannot create a host yourself.
 *
 * The best way to get an host instance is to call the static method 
 * Host.getByName().
 *
 * For example to get the instance of the host. If your platform
 * file description contains an host named "Jacquelin" :
 *
 * \verbatim
Host jacquelin;

try { 
	jacquelin = Host.getByName("Jacquelin");
} catch(HostNotFoundException e) {
	System.err.println(e.toString());
}
...
\endverbatim
 *
 */ 
public class Host {

	/**
	 * This attribute represents a bind between a java host object and
	 * a native host. Even if this attribute is public you must never
	 * access to it. It is set automatically during the call of the 
	 * static method Host.getByName().
	 *
	 * @see				Host.getByName().
	 */ 
	public long bind;


	/**
	 * User data.
	 */ 
	private Object data;
	protected Host() {
		this.bind = 0;
		this.data = null;
	};

	/**
	 * This static method gets an host instance associated with a native
	 * host of your platform. This is the best way to get a java host object.
	 *
	 * @param name		The name of the host to get.
	 *
	 * @exception		HostNotFoundException if the name of the host is not valid.
	 *					NativeException if the native version of this method failed.
	 */ 
	public static Host getByName(String name) 
	throws HostNotFoundException {
		if (name==null)
			throw new NullPointerException("No host can have a null name");
		return MsgNative.hostGetByName(name);
	}

	/**
	 * This static method returns the count of the installed hosts.
	 *
	 * @return			The count of the installed hosts.
	 *
	 */ 
	public static int getCount() {
		return MsgNative.hostGetCount();
	}

	/**
	 * This static method return an instance to the host of the current process.
	 *
	 * @return			The host on which the current process is executed.
	 */ 
	public static Host currentHost() {
		return MsgNative.hostSelf();
	}

	/**
	 * This static method returns all of the hosts of the installed platform.
	 *
	 * @return			An array containing all the hosts installed.
	 *
	 */ 
	public static Host[] all()  {
		return MsgNative.allHosts();
	}

	/**
	 * This method returns the name of a host.
	 *
	 * @return			The name of the host.
	 *
	 */ 
	public String getName()  {
		return MsgNative.hostGetName(this);
	}

	/**
	 * Sets the data of the host.
	 *
	 */ 
	public void setData(Object data) {
		this.data = data;
	} 
	/**
	 * Gets the data of the host.
	 */ 
	public Object getData() {
		return this.data;
	}

	/**
	 * Checks whether a host has data.
	 */ 
	public boolean hasData() {
		return null != this.data;
	}

	/**
	 * This method returns the number of tasks currently running on a host.
	 * The external load is not taken in account.
	 *
	 * @return			The number of tasks currently running on a host.
	 */ 
	public int getLoad() {
		return MsgNative.hostGetLoad(this);
	}

	/**
	 * This method returns the speed of the processor of a host,
	 * regardless of the current load of the machine.
	 *
	 * @return			The speed of the processor of the host in flops.
	 *
	 */ 
	public double getSpeed() {
		return MsgNative.hostGetSpeed(this);
	}

	/** This method tests if a host is avail. */
	public boolean isAvail() {
		return MsgNative.hostIsAvail(this);
	}
} 
